src/libstd/panicking.rs

```diff
@@ -466,20 +483,11 @@ fn rust_panic_with_hook(payload: &mut dyn BoxMeUp,
             Location::internal_constructor(file, line, col),
         );
         HOOK_LOCK.read();
+        info.set_payload(payload.get());
         match HOOK {
-            // Some platforms know that printing to stderr won't ever actually
-            // print anything, and if that's the case we can skip the default
-            // hook.
-            Hook::Default if stderr_prints_nothing() => {}
```

Comment:
I made sure the changes should have the same functionality, and I also checked that `set_payload` isn't really doing much so it's ok to do unconditionally. But yeah I suppose some form of LTO might no longer work.

The crux is that currently in this target, checking whether there is a panic output mechanism prevents future use of the same (it's an atomic “take”). I can think if there's another way to restructure all this tomorrow, but this was the best I could come up with a few months ago :)

Reply:
Sorry, with regards to this and https://github.com/rust-lang/rust/pull/56066#discussion_r235098170, I conflated two things a little bit.

In the SGX target, whether panicking can ever output anything at all is determined post-link time by setting/unsetting the `DEBUG` global flag. This is because panic message may contain sensitive information, so it's opt-in. This requirement leads to the function signature: `fn panic_output() -> Option<PanicOutput>`. Ideally having `DEBUG = false` leads to the same fast-path as wasm without a panic implementation.

The second requirement that leads to the "writer factory" in `PanicOutput::Custom` is that in this target the panic writer can only be constructed once.

Reply:
I refactored this a bunch to use types instead of bare functions. I think it should be LTOable. Please take a look.